generators/commons/src/ast/AbstractWriter.ts

```diff
@@ -0,0 +1,161 @@
+import { AbstractAstNode } from "./AbstractAstNode";
+import { CodeBlock } from "./CodeBlock";
+
+const TAB_SIZE = 4;
+
+export class AbstractWriter {
+    /* The contents being written */
+    public buffer = "";
+    /* Indentation level (multiple of 4) */
+    private indentLevel = 0;
+    /* Whether anything has been written to the buffer */
+    private hasWrittenAnything = false;
+    /* Whether the last character written was a newline */
+    private lastCharacterIsNewline = false;
+
+    /**
+     * Writes arbitrary text
+     * @param text
+     */
+    public write(text: string): void {
+        const textEndsInNewline = text.length > 0 && text.endsWith("\n");
+        // temporarily remove the trailing newline, since we don't want to add the indent prefix after it
+        const textWithoutNewline = textEndsInNewline ? text.substring(0, text.length - 1) : text;
+
+        const indent = this.getIndentString();
+        let indentedText = textWithoutNewline.replaceAll("\n", `\n${indent}`);
+        if (this.isAtStartOfLine()) {
+            indentedText = indent + indentedText;
+        }
+        if (textEndsInNewline) {
+            indentedText += "\n";
+        }
+        this.writeInternal(indentedText);
+    }
+
+    /**
+     * Writes a node
+     * @param node
+     */
+    public writeNode(node: AbstractAstNode): void {
+        node.write(this);
+    }
+
+    /**
+     * Writes a node but then suffixes with a `;` and new line
+     * @param node
+     */
+    public writeNodeStatement(node: AbstractAstNode): void {
+        node.write(this);
+        this.write(";");
```

Comment:
Details like this might need to be left out of the abstract classes - not every language ends statements with a `;`. I know most languages do use this syntax though, so we have a couple options:

1. Leave this as-is and rely on method overrides in language-specific writers.
2. Refactor the `AbstractWriter` to be fairly similar to the original `generators/commons.Writer`.

You might have already had (1) as the desired solution, but figured it was worth calling out.